fs-api/src/System/FS/API.hs

```diff
@@ -175,3 +188,188 @@ hClose' HasFS { hClose, hIsOpen } h = do
+-- but is separated for now so downstream code does not break.
+data HasBufFS m h = HasBufFS {
+    -- | Like 'hGetSome', but the bytes are read into a user-supplied buffer.
+    -- See __User-supplied buffers__.
+    hGetBufSome      :: HasCallStack
+                     => Handle h
+                     -> MutableByteArray (PrimState m) -- ^ Buffer to read bytes into
+                     -> BufferOffset -- ^ Offset into buffer
+                     -> ByteCount -- ^ The number of bytes to read
+                     -> m ByteCount
+    -- | Like 'hGetSomeAt', but the bytes are read into a user-supplied buffer.
+    -- See __User-supplied buffers__.
+  , hGetBufSomeAt    :: HasCallStack
+                     => Handle h
+                     -> MutableByteArray (PrimState m)   -- ^ Buffer to read bytes into
+                     -> BufferOffset -- ^ Offset into buffer
+                     -> ByteCount -- ^ The number of bytes to read
+                     -> AbsOffset -- ^ The file offset at which to read
+                     -> m ByteCount
+    -- | Like 'hPutSome', but the bytes are written from a user-supplied buffer.
+    -- See __User-supplied buffers__.
+  , hPutBufSome      :: HasCallStack
+                     => Handle h
+                     -> MutableByteArray (PrimState m) -- ^ Buffer to write bytes from
+                     -> BufferOffset -- ^ Offset into buffer
+                     -> ByteCount -- ^ The number of bytes to write
+                     -> m ByteCount
+    -- | Like 'hPutSome', but the bytes are written from a user-supplied buffer
+    -- at a given file offset. This offset does not affect the offset stored in
+    -- the file handle (see also 'hGetSomeAt'). See __User-supplied buffers__.
+  , hPutBufSomeAt    :: HasCallStack
+                     => Handle h
+                     -> MutableByteArray (PrimState m)   -- ^ Buffer to write bytes from
+                     -> BufferOffset -- ^ Offset into buffer
+                     -> ByteCount -- ^ The number of bytes to write
+                     -> AbsOffset -- ^ The file offset at which to write
+                     -> m ByteCount
+  }
+
+-- | Wrapper for 'hGetBufSomeAt' that ensures that we read all bytes from a
+-- file.
+--
+-- A sufficiently large buffer can be provided by comparing 'hGetSize' against
+-- the requested file offset.
+--
+-- Is implemented in terms of 'hGetBufExactlyAt'.
+hGetBufAllAt :: forall m h. (HasCallStack, MonadThrow m)
+          => HasFS m h
+          -> HasBufFS m h
+          -> Handle h
+          -> MutableByteArray (PrimState m)   -- ^ Buffer to read bytes into
+          -> BufferOffset -- ^ Offset into buffer
+          -> AbsOffset -- ^ The file offset at which to read
+          -> m ByteCount
+hGetBufAllAt hfs hbfs h buf bufOff off = do
+    sz <- hGetSize hfs h
+    let c = sz - unAbsOffset off
+    hGetBufExactlyAt hfs hbfs h buf bufOff (fromIntegral c) off
```

Comment:
I think I'd omit this one. I know there's the corresponding lazy bytestring one, but here it's a bit dubious because the caller needs to already know the size of the file (or an upper bound on it) to have allocated a large enough buffer.

Given that it's fairly simple to implement in terms of `hGetBufExactly` then I'd just leave it to applications in case they need it.

And in practice, if you're going to read the whole file, it'll be the whole file from the beginning, not from an offset. Too niche.

Reply:
> I think I'd omit this one. I know there's the corresponding lazy bytestring one, but here it's a bit dubious because the caller needs to already know the size of the file (or an upper bound on it) to have allocated a large enough buffer.

Good point, oversight from me

> Given that it's fairly simple to implement in terms of `hGetBufExactly` then I'd just leave it to applications in case they need it.
>
> And in practice, if you're going to read the whole file, it'll be the whole file from the beginning, not from an offset. Too niche.

:+1: